Azure/azure-cli
# {Core} Reducing the risk of logging file fd leaking

**Description**  
As I discovered in https://github.com/Azure/azure-cli/pull/12949, CLI would crash after long time running due to running out of available fd because it didn't close the fd properly.

Running tests with pytest, testing is terminated after the number of avaliable fd is consumed.
![image](https://user-images.githubusercontent.com/14357159/78963857-65cd8780-7b2b-11ea-8ec5-5a2779920c9f.png)
![image](https://user-images.githubusercontent.com/14357159/78966209-c2cc3c00-7b31-11ea-9fd2-a4c8d7a83f8f.png)

I add a [hook `doCleanups` of unittest.Testcase](https://docs.python.org/3/library/unittest.html#unittest.TestCase.doCleanups) to mandatorily clean those unclosed fd hold by CLI's logger. 

**This could reduce the risk of fd leaking but still need developer to manually call `end_cmd_metadata_logging()` of AzLogging, like:**
https://github.com/Azure/azure-cli/blob/806e4318d058c3037566fa5a7a07cfc99d1ff51d/src/azure-cli/azure/cli/__main__.py#L49-L50

**Why I say reducing**
1.  there are still some test classes didn't inherit from `ScenarioTest/LiveScenarioTest` but `unittest.TestCase`, so, they couldn't benifit from this fix.
2.  A log files rotated logic is applied to delete older log files to keep at most 25 log files in order to save the disk. If there are too many commands/tests running at the same time and CLI instance can't recover by Python Garbage Collector, it will keep leak the fd until CLI instance is del or the process is terminated. Like these 2 PRs: https://github.com/Azure/azure-cli/issues/12882, https://github.com/Azure/azure-cli/issues/10435

In both scenarios, we will still have chance to see `(deleted)` files that are unclosed and leaking

**Testing Guide**  
1.  uncomment the new added lines and run a test with lots of test units like `test_network_commands.py`,
2.  what over the opened fd by your process, `lsof -p {TestProcessID} | grep "azure"`,
3.  After a while, you will see lots of items end up with `(deleted)` like the image above. That's the leaking fds. 
4.  If you run with pytest `pytest -x -v --capture=no` and with tons of tests, process will crash.
5.  Adding the new added line in this PR, there are still sometime that the **deleted** file will show in `lsof -p {TestProcessID}`, but they will disappear very soon.

**History Notes**  



---

This checklist is used to make sure that common guidelines for a pull request are followed.
